tests/python/tests.py

```diff
@@ -1,22 +1,158 @@
+from typing import Tuple
 import psycopg2
+import psutil
+import os
+import signal
+import subprocess
+from threading import Thread
+import time
 
-def test_normal_db_access():
-    conn = psycopg2.connect("postgres://sharding_user:sharding_user@127.0.0.1:6432/sharded_db?application_name=testing_pgcat")
+SHUTDOWN_TIMEOUT = 5
+
+PGCAT_HOST = "127.0.0.1"
+PGCAT_PORT = "6432"
+
+
+def pgcat_start():
+    pg_cat_send_signal(signal.SIGTERM)
+    pgcat_start_command = "./target/debug/pgcat .circleci/pgcat.toml"
+    subprocess.Popen(pgcat_start_command.split())
+
+
+def pg_cat_send_signal(signal: signal.Signals):
+    for proc in psutil.process_iter(["pid", "name"]):
+        if "pgcat" == proc.name():
+            os.kill(proc.pid, signal)
+
+
+def connect_normal_db(
+    autocommit: bool = False,
+) -> Tuple[psycopg2.extensions.connection, psycopg2.extensions.cursor]:
+    conn = psycopg2.connect(
+        f"postgres://sharding_user:sharding_user@{PGCAT_HOST}:{PGCAT_PORT}/sharded_db?application_name=testing_pgcat"
+    )
+    conn.autocommit = autocommit
     cur = conn.cursor()
 
+    return (conn, cur)
+
+
+def cleanup_conn(conn: psycopg2.extensions.connection, cur: psycopg2.extensions.cursor):
+    cur.close()
+    conn.close()
+
+
+def test_normal_db_access():
+    conn, cur = connect_normal_db()
     cur.execute("SELECT 1")
     res = cur.fetchall()
     print(res)
+    cleanup_conn(conn, cur)
 
 
 def test_admin_db_access():
-    conn = psycopg2.connect("postgres://admin_user:admin_pass@127.0.0.1:6432/pgcat")
-    conn.autocommit = True # BEGIN/COMMIT is not supported by admin db
+    conn = psycopg2.connect(
+        f"postgres://admin_user:admin_pass@{PGCAT_HOST}:{PGCAT_PORT}/pgcat"
+    )
+    conn.autocommit = True  # BEGIN/COMMIT is not supported by admin db
     cur = conn.cursor()
 
     cur.execute("SHOW POOLS")
     res = cur.fetchall()
     print(res)
+    cleanup_conn(conn, cur)
+
+
+def test_shutdown_logic():
```

Comment:
We use bash for these kind of tests, why Python?

Reply:
Since we're operating on the same postgres connection and doing things in between queries and transactions it felt easiest to do it in python.